### DISCUSSION TITLE
How to add a dependency to all `POST` API endpoints?

### DISCUSSION THREAD

**Discussion by cyberksh** [fastapi/fastapi#6033] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import fastapi
import fastapi.routing
import pydantic

app = fastapi.FastAPI()


class Item(pydantic.BaseModel):
    name: str
    price: int


router = fastapi.APIRouter()


@router.get("/item")
async def get_item() -> list[Item]:
    return [Item(name="test item", price=2)]


async def verify_key(x_key: str = fastapi.Header()):
    if x_key != "super_secret_key":
        raise fastapi.HTTPException(status_code=400, detail="X-Key header invalid")


@router.post("/item")
def create_item(item: Item):
    return item


# Hacky way to inject a dependency on all `POST` calls
for route in router.routes:
    assert isinstance(route, fastapi.routing.APIRoute)
    if "POST" in route.methods:
        route.dependencies.append(fastapi.Depends(verify_key))

app.include_router(router)
```

### Description

I am trying to add a dependency to all POST calls present in the API.
- There are multiple `POST` calls spread across multiple modules in the API. 
- I don't want to add it to each `POST` endpoint, neither do I want to add it to the `APIRouter` because a `router` can contain both `GET` and `POST` calls.
- I could get it working with the above example. 

Is there a better way to do this?

### Operating System

Linux

### FastAPI Version

0.92.0

### Python Version

Python 3.10.9

**Comment by ebottos94** [comment, marked as answer]
Hi @cyberksh you can do something like this : 

in ```your_router.py``` : 

```python
from fastapi import APIRouter

router = APIRouter(
    prefix='/api', 
    tags=['api']
    )

@router.get("/")
async def get_function():
    return "hello world from get route"

@router.post("/")
async def post_function():
    return "hello world from post route"
```

in ```main.py``` : 

```python
from fastapi import FastAPI, Header, Request, Depends, HTTPException
from .your_router import router


def verify_key(x_key: str = Header()):
    if x_key != "super_secret_key":
        raise HTTPException(status_code=400, detail="X-Key header invalid")

def check_dependencies(request: Request, apy_key=None) :
    if request.method == 'POST' :
        if not apy_key :
            raise HTTPException(status_code=400, detail="X-Key header invalid")
        else :
            verify_key(apy_key)

app = FastAPI()

app.include_router(router, dependencies=[Depends(check_dependencies)])
```

and repeat this in all your routers.

**Reply by cyberksh** [reply]
Thanks for the response but this doesn't work for my usecase. The Swagger page generated from the code doesn't have the field to set the header. It was never added to the OpenAPI spec.

**Reply by ebottos94** [reply]
If you integrated the code correctly, Depends injects "apy_key" value in your routes in Swagger interface. Post your code if you have other doubts. Anyway, you can test your api with postman also.

**Reply by cyberksh** [reply]
Yeah there was a mistake in my code. It does add the header field in the Swagger page. It adds it to even the `GET` requests but the validation is never executed. I guess this solves my issue. Thanks.
